### PR TITLE
[develop] Trello-HfkQfdwG. Password recovery

### PR DESCRIPTION
**Pull request by AndreySurzhan**
[Password recovery ticket
](https://trello.com/c/HfkQfdwG)
- [x] POST Method and route for requesting pass recovery and generating pass recovery token
- [x] Sending email with generated url with username and token
- [x] GET Method and route for requesting password changing
- [x] POST Method and route for password changing

-----------------------
Full workflow descriptions [Password recovery page](https://github.com/wildeastengineer/MKeepAPI/wiki/Password-recovery)

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -12,5 +12,7 @@
             "clientId": "MKWebAppV1",
             "clientSecret": "mbO2FdS451lEz0fM8FlSM3n1rbokBSyy"
         }
-    ]
+    ],
+    "host": "localhost",
+    "protocol": "http"
 }
```
Comment (wildeastengineer) [inline config.json]: Empty line in the end of file. I know that I missed it.

File: controllers/user.js [hunk 2]
```diff
@@ -49,6 +55,67 @@ var userController = {
...
+
+                passRecoveryUrl = url.format({
+                    protocol: config.get('protocol') || 'http',
+                    hostname: config.get('host'),
+                    port: config.get('port'),
+                    query: {
+                        username: user.username,
+                        token: user.passRecoveryToken
+                    }
+
```
Comment (wildeastengineer) [inline controllers/user.js]: extra line

File: controllers/user.js [hunk 3]
```diff
@@ -49,6 +55,67 @@ var userController = {
...
+                        .then(function () {
+                            deferred.resolve(user);
+                        });
+                });
+            }
+        );
+
+        return deferred.promise;
+
+        function generateRecoveryToken (user) {
```
Comment (wildeastengineer) [inline controllers/user.js]: You use this function only in one place and it cam be written in one line. Do you need to exclude this code to separate function?

File: utils/emailSender.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        user: 'money.keeper.service@gmail.com',
+        pass: 'moneyKeeper123456'
+    }
+};
+
+var emailSender = {
+    passwordRecovery: function (userEmail, url) {
+        var deferred = Q.defer();
+        var transporter = nodemailer.createTransport(smtpConfig);
+
```
Comment (wildeastengineer) [inline utils/emailSender.js]: extra line

File: utils/emailSender.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            to: userEmail,
+            subject: 'Password recovery for you MKeeper account', // Subject line
+            text: url, // plaintext body
+            html: '<b>'+ url + '</b>' // html body
+        };
+
+        // send mail with defined transport object
+        transporter.sendMail(mailOptions, function(error, info){
+            if(error){
+                console.log(error);
```
Comment (wildeastengineer) [inline utils/emailSender.js]: use logger

File: utils/emailSender.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            },
+            to: userEmail,
+            subject: 'Password recovery for you MKeeper account', // Subject line
+            text: url, // plaintext body
+            html: '<b>'+ url + '</b>' // html body
+        };
+
+        // send mail with defined transport object
+        transporter.sendMail(mailOptions, function(error, info){
+            if(error){
```
Comment (wildeastengineer) [inline utils/emailSender.js]: cs

File: utils/emailSender.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            console.log('Message sent: ' + info.response);
+
+            deferred.resolve(info);
+        });
+
+        return deferred.promise;
+    }
+};
+
+module.exports = emailSender;
```
Comment (wildeastengineer) [inline utils/emailSender.js]: extra line

File: controllers/user.js [hunk 8]
```diff
@@ -97,6 +97,7 @@ var userController = {
         return deferred.promise;
 
         /**
+         * @privet
```
Comment (wildeastengineer) [inline controllers/user.js]: Привет :smile:

**Comment by AndreySurzhan** [comment]
@wildeastengineer could you please review

File: controllers/user.js [hunk 9]
```diff
@@ -24,33 +27,240 @@ var userController = {
...
         return deferred.promise;
+    },
+    /**
+     * Generate password recovery token
+     *
+     * @param {string} username - email in fact
+     *
+     * @returns {promise}
+     */
+    generatePassRecoveryToken: function (username) {
```
Comment (wildeastengineer) [inline controllers/user.js]: Maybe it's better to storerecovery tokens in separate collection (like access and refresh tokens) and keep user model clear?

File: utils/templates/passwordRecovery.html [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<p>
+    Dear,<br>
+    {username}.
+</p>
+<p>
+    If you want to recover you password please go to <a href="{link}">password recovery link</a>.
```
Comment (wildeastengineer) [inline utils/templates/passwordRecovery.html]: to recover you**r** password

File: utils/emailSender.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        var deferred = Q.defer();
+        var transporter = nodemailer.createTransport(smtpConfig);
+
+        var mailOptions = {
+            from: {
+                name: 'MKeeper Service',
+                address: smtpConfig.auth.user
+            },
+            to: userEmail,
+            subject: 'Password recovery for your MKeeper account', // Subject line
```
Comment (wildeastengineer) [inline utils/emailSender.js]: I think it would be better if you use any translator here ([i18n](https://github.com/mashpie/i18n-node) for example)

Comment (AndreySurzhan) [inline utils/emailSender.js]: Do we really need a translation on backend side?
I mean that is it really important for client side.
If you still think we need it here lets implement it in next iteration then, ok?

File: controllers/user.js (hunk 9 above)
Comment (AndreySurzhan) [inline controllers/user.js]: There are some recommendations on how DB should be designed:
[tutorialspoint](http://www.tutorialspoint.com/mongodb/mongodb_data_modeling.htm)
[stackoverfow](http://stackoverflow.com/questions/5373198/mongodb-relationships-embed-or-reference)
In few words: put as much as possible in one document - it makes search through DB much efficient due to DB features

Comment (AndreySurzhan) [inline controllers/user.js]: WDYT @wildeastengineer ?

Comment (wildeastengineer) [inline controllers/user.js]: @AndreySurzhan I think we have already started writing token in their own collections. And recovery password will be happening infrequently. So I would not bother about performance in this case.

**Comment by AndreySurzhan** [comment]
@wildeastengineer Could you please review.
I rewrite almost all pass recovery related methods and put password recovery token into separate model and created separate controller for that

File: utils/emailSender.js (hunk 11 above)
Comment (wildeastengineer) [inline utils/emailSender.js]: @AndreySurzhan it's not really backend part. It is message for user as I understand.

File: controllers/passRecovery.js [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+                        logger.info('New password recovery token has been successfully created: ' + token);
+                        deferred.resolve(token);
+                    }
+                });
+            }
+        );
+
+        return deferred.promise;
+    },
+    /**
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: extra space

File: controllers/passRecovery.js [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                }
+
+                logger.info(result.message + ': ' + token._id) ;
+                deferred.resolve(result);
+            }
+        );
+
+        return deferred.promise;
+    },
+    /**
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: extra space

File: controllers/passRecovery.js [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+     * @param {string} token - password recovery token that is passed as param in url
+     *
+     * @returns {promise}
+     */
+    removePassRecoveryToken: function (token) {
+        var deferred = Q.defer();
+
+        PassRecoveryTokenModel.findOneAndRemove(
+            {
+            token: token
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: tab

File: controllers/passRecovery.js [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        return deferred.promise;
+    },
+    /**
+     * Check whether given token does exist in data base
+     *
+     * @param {string} token - password recovery token that is passed as param in url
+     *
+     * @returns {promise}
+     */
+    exists: function (token) {
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: Maybe add to method name what exactly exists?

File: controllers/passRecovery.js [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+                }
+
+                if (!token) {
+                    logger.error('Password recovery token doesn\'t exist');
+                    deferred.reject(new Error('Password recovery token doesn\'t exist'));
+
+                    return;
+                }
+
+                logger.info(result.message + ': ' + token._id) ;
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: extra space before `;`

File: routes/user.js [hunk 17]
```diff
@@ -1,9 +1,16 @@
 /// Libs
+var config = require('../libs/config');
+var emailSender = require('../utils/emailSender');
 var oauth2 = require('../libs/oauth2');
+var url = require('url');
 /// Controllers
 var userController = require('../controllers/user.js');
+var passRecovery = require('../controllers/passRecovery.js');
+
```
Comment (wildeastengineer) [inline routes/user.js]: extra line

File: routes/user.js [hunk 18]
```diff
@@ -27,6 +34,59 @@ var userRegisterRoutes = function (router, authenticate, sendError) {
...
+                passRecoveryUrl = url.format({
+                    protocol: config.get('protocol') || 'http',
+                    hostname: config.get('host'),
+                    pathname: config.get('baseUrl') + recoverPasswordPath,
+                    port: config.get('port'),
+                    query: {
+                        username: req.body.username,
+                        token: token.token
+                    }
+
```
Comment (wildeastengineer) [inline routes/user.js]: extra line

File: utils/emailSender.js [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    passwordRecovery: function (userEmail, url) {
+        var deferred = Q.defer();
+        var result = {
+            success: true,
+            message: 'Password recovery token has been successfully generated. ' +
+                    'Notification email has been sent to user'
+        };
+        var transporter = nodemailer.createTransport(smtpConfig);
+
```
Comment (wildeastengineer) [inline utils/emailSender.js]: extra line

File: utils/emailSender.js (hunk 11 above)
Comment (wildeastengineer) [inline utils/emailSender.js]: @AndreySurzhan but we can postpone this issue to future iteration

Comment (AndreySurzhan) [inline utils/emailSender.js]: @wildeastengineer I created separate ticket for that tast
https://trello.com/c/dCBOKoCk

**Comment by AndreySurzhan** [comment]
@wildeastengineer PR has been updated. Could you please review?

File: controllers/passRecovery.js [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+            },
+            function (error, user) {
+                var newPassRecoveryToken;
+                var token;
+
+                if (error) {
+                    logger.error(error);
+                    deferred.reject(error);
+
+                    return;
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: `return deferred.promise;`

File: controllers/passRecovery.js [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+                }
+
+                if (!user) {
+                    logger.error('Cannot create password recovery token. ' +
+                            'User with this email doesn\'t exist' + username);
+                    deferred.reject(new Error('Cannot create password recovery token. ' +
+                            'User with this email doesn\'t exist' + username));
+
+                    return;
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: `return deferred.promise;`

File: controllers/passRecovery.js [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+        PassRecoveryTokenModel.findOne(
+            {
+                token: token
+            },
+            function (error, token) {
+                if (error) {
+                    logger.error(error);
+                    deferred.reject(error);
+
+                    return;
```
Comment (wildeastengineer) [inline controllers/passRecovery.js]: `return deferred.promise;`

**Comment by wildeastengineer** [comment]
:+1:

**Pull request merged**